Focal Operations of Immediate Neighbours

fosskers opened the pull request:
### TODO

- [x] FocalClassification
- [x] FocalLinkage, FocalLength (lineal conditions)
- [x] FocalPartition, FocalFrontage, FocalArea (areal conditions)
- [x] FocalVolume, FocalGradient, FocalAspect, FocalDrainage (surficial conditions)

**Omitted:** 
- FocalCombination (not useful as its own operation)
- FocalRanking (how is this different from FocalVariety?)
- FocalInsularity (unclear how it would operate in parallel. Seems to require `State`-like memory, hence it's monadic, hence it's serial)
- The variant of FocalFrontage and FocalArea that allow for also supplying an elevation layer

fosskers committed:
fclassify

fosskers commented:
Hey @lehins , looking at [Border](https://hackage.haskell.org/package/massiv-0.1.4.0/docs/Data-Massiv-Core-Index.html#t:Border) would it be reasonable to add an extra value to this sum type that takes a function? Say `Depends (a -> b)`? 

For focal operations over immediate neighbourhoods (1 extra pixel around the focus), `Fill` acts as `const`, `Edge` and `Reflect` act as `id`, and `Wrap` and `Continue` act as custom functions. I'd *really* like to be able to inspect the focus manually and set it to whatever I'd like. Specifically, I have a case where I'd love to set the border pixels with `Depends succ` so that I know they will be detected as different than the focus.

fosskers commented:
Crap... `Border` only has one type variable. Okay then... `Depends (a -> a)` ? :)

lehins commented:
Not only that, what would be the argument `a` in `Depends (a -> b)`? What would it actually depend on?

lehins commented:
Maybe I am not understanding what you mean, could you give me a code example? Possibly that focal operation?

fosskers commented:
The argument would be whatever the neighbourhood centre would be, different for every time the off-the-edge values are being calculated. 

Phrased differently: If `Fill a` sets any off-the-Array cell to `a`, then `Depends (a -> a)` sets off-the-Array cells to the result of some given function applied to the edge.

The real scenario is that I'm implementing an operation called `Focal Linkage`:
```haskell
-- | Focal Linkage - a description of how each neighbourhood focus is connected
-- to its neighbours. Foci of equal value to none of their neighbours will have
-- an empty `S.Set`.
flinkage :: (Default a, Eq a, Manifest u Ix2 a) => 
  Raster u p r c a -> Raster DW p r c (S.Set Direction)
flinkage (Raster a) = Raster $ mapStencil linkStencil a

-- | Directions that neighbourhood foci can be connected by. See `flinkage`
-- and `flength`.
data Direction = East | NorthEast | North | NorthWest | West | SouthWest | South | SouthEast
  deriving (Eq, Ord, Show)

linkStencil :: (Default a, Eq a) => Stencil Ix2 a (S.Set Direction)
linkStencil = ...
```
"Connected" means sharing the same value as the neighbourhood focus (hence `Eq`). Edge pixels can't possibly be connected to things off the edge of the Array itself, so for pixel `(0, 0)`, when it checks for the value at `(-1 :. -1)`, I need that to be some *different* value than the focus, so that equality will fail.

Currently I'm using `Continue` since it wraps but doesn't repeat the edge value. Unfortunately, this isn't correct for areas around the edges that all share the same value. An array like:
```
[ [ 1, 1 ]
, [ 1, 1 ] ]
```
would report mistakenly that each cell was connected to outside cells. `(0, 0)` would mistakenly contain `North`, `NorthWest`, and `West` in its `Set Direction` in this case.

All the current `Border` values (`Fill`, etc) run the risk of the yielding values equal to the focus, which would result in a mistaken connection.

fosskers committed:
flinkage

lehins commented:
@fosskers I see what you mean now. `Border` resolution isn't stencil specific in any way, it shows up in other places, not only in `massiv`, but it also in [hip](https://github.com/lehins/hip), the library that actually motivated my work on `massiv`. It shows up in interpolation, rotation, and in all sorts of other places, even indexing functions are implemented using `handleBorderIndex` function. Reason why I am going in all this in such detail is because there isn't necessarily gonna a be a notion of center and even any other available value in places where `Border` is used, besides the offending index. I think, it would be possible to add another type constructor to `Border e`, such that `Generate :: forall ix . Index ix => ix -> e`, which would produce an element from a faulty index, but I don't think it be too helpful in your problem. I'll think a bit more on this.

Now, having `Border` aside, it is not a strict requirement to use `Stencil` in order to get all of it's performance benefits. You can always construct a `DW` image type manually using [`makeWindowedArray`](https://www.stackage.org/haddock/lts-11.4/massiv-0.1.4.0/Data-Massiv-Array-Delayed.html#v:makeWindowedArray) function. It is a bit trickier though, that's why `Stencil` actually hides that bit of ugliness.

Short description of `makeWindowedArray`: create a delayed image that will handle elements at the border, pass it as a first argument, similarly for the window function, but it is safe to assume that it won't be called near the border (depends on the startf index and window size), so you can use `unsafeIndex` function in there as long is it is not gonna be indexing outside of that window, so it is totally safe :) `Stencil` enforces some safety on the user, which can be bypassed when using `unsafe*` functions.

fosskers commented:
For now, for `flinkage` only, I'll rely on being a little less robust in favour of simplicity.

lehins commented:
@fosskers Got some good news for you. I decided to implement an unsafe version of stencil, just for those that know what they are doing and really need one. Here is [`forStencilUnsafe`](https://github.com/lehins/massiv/commit/d28ef186f4164662d53a3d373ac38f809de2d263#diff-71778d171f68f3a837076d58bca58bd8R231), which is sort of like `makeStencil` composed with `mapStencil`, except that it is just as fast but unsafe. It will not try to be to smart around the border it's up to the user. Here is a usage example:

```
average3x3FilterUnsafe :: (Manifest r Ix2 a, Fractional a) => Array r Ix2 a -> Array DW Ix2 a
average3x3FilterUnsafe arr = forStencilUnsafe arr (3 :. 3) (1 :. 1) $ \ get ->
  let get' = fromMaybe 0 . get in
  (  get' (-1 :. -1) + get' (-1 :. 0) + get' (-1 :. 1) +
     get' ( 0 :. -1) + get' ( 0 :. 0) + get' ( 0 :. 1) +
     get' ( 1 :. -1) + get' ( 1 :. 0) + get' ( 1 :. 1)   ) / 9
{-# INLINE average3x3FilterUnsafe #-}
```

Let me know how it works out for you. I'll hold off on making a release, will wait and see if there is some other feature you can come up with ;)

fosskers committed:
flength

fosskers committed:
fpartition

fosskers committed:
Fixed `fpartition`

fosskers committed:
Generalize corner checking

fosskers committed:
fareals

- With some corrections to the behaviour of `surround` and the
  `fpartition` stencil.

fosskers committed:
ffrontage

fosskers committed:
farea

fosskers committed:
Bump massiv versions

fosskers committed:
fvolume

fosskers committed:
leftPseudo

fosskers committed:
fgradient

fosskers committed:
faspect

- Can fail (w/ Nothing) if the pixel is flat!

fosskers committed:
Make `grayscale`, `display`, and `classify` more generic

fosskers committed:
Fix `fgradient` and `faspect`

- And also make them more generic. If any of these functions output
  `Double`, the input should be some `Real` that can be converted.
  Instead of forcing the user to do a manual `fmap` and re-strict.

fosskers committed:
Improve `faspect` docs

fosskers committed:
Bump stack resolver

fosskers committed:
fdownstream

fosskers committed:
Fix issue w/ pits in fdownstream

fosskers committed:
Improve `fdownstream` performance

fosskers committed:
fdownstream'

- `Word`-based, as described on Page 81 of GaCM

fosskers committed:
Must more efficient `fdownstream` and `fupstream`
